pkg/dns/service.go

```diff
@@ -224,28 +225,46 @@ func (s *Service) SetEndpoints(ctx context.Context, addresses []gatewayv1beta1.G
 
 // GetManagedZoneForHost returns a ManagedZone and subDomain for the given host if one exists.
 //
-// Currently, this returns the first matching ManagedZone found in the traffic resources own namespace, or if none is found,
-// it looks for the first matching one listed in the default ctrl namespace.
+// Currently, this returns the first matching ManagedZone found in the traffic resources own namespace
 func (s *Service) GetManagedZoneForHost(ctx context.Context, host string, t traffic.Interface) (*v1alpha1.ManagedZone, string, error) {
+	var managedZones v1alpha1.ManagedZoneList
+	if err := s.controlClient.List(ctx, &managedZones, client.InNamespace(t.GetNamespace())); err != nil {
+		log.FromContext(ctx).Error(err, "unable to list managed zones in traffic resource NS")
+		return nil, "", err
+	}
+	return FindMatchingManagedZone(host, host, managedZones.Items)
+}
+
+func FindMatchingManagedZone(originalHost, host string, zones []v1alpha1.ManagedZone) (*v1alpha1.ManagedZone, string, error) {
+	if len(zones) == 0 {
+		return nil, "", fmt.Errorf("no managed zone found for host : %s", host)
+	}
+	host = strings.ToLower(host)
 	hostParts := strings.SplitN(host, ".", 2)
 	if len(hostParts) < 2 {
-		return nil, "", fmt.Errorf("unable to parse host : %s on traffic resource : %s", host, t.GetName())
+		return nil, "", fmt.Errorf("unable to parse host : %s", host)
 	}
-	subDomain := hostParts[0]
-	parentDomain := hostParts[1]
 
-	var managedZones v1alpha1.ManagedZoneList
+	//get the TLD from this host
+	tld, icann := publicsuffix.PublicSuffix(host)
 
-	if err := s.controlClient.List(ctx, &managedZones, client.InNamespace(t.GetNamespace()), client.MatchingFields{"spec.domainName": parentDomain}); err != nil {
-		log.FromContext(ctx).Error(err, "unable to list managed zones in traffic resource NS")
-		return nil, "", err
+	//The host is just the TLD, or the detected TLD is not an ICANN TLD
+	if host == tld || !icann {
```

Comment:
Whats the significance of it being icann or not? Would that affect private networks using some internal dns names?

Reply:
> Whats the significance of it being icann or not? Would that affect private networks using some internal dns names?

yeah, ICANN is a boolean that is false if the TLD is not managed by ICANN (e.g. `.phil`)

Reply:
> > Whats the significance of it being icann or not? Would that affect private networks using some internal dns names?
> 
> yeah, ICANN is a boolean that is false if the TLD is not managed by ICANN (e.g. `.phil`)

I'm not against allowing non-ICANN domains, I just assumed it was safer to default to not allowing them.

Reply:
Yeah I'm not sure tbh. I guess we can just go with it, but we should probably prevent the creation of ManagedZones that are using a domain that isn't icann managed in that case. Maybe a separate issue that needs discussion though. One thing i had thought was if we to implement a CoreDNS dns provider for local dev we could use a `foo.local` domain or something which a assume this wouldn't allow.

Reply:
ok, I'll allow non-ICANN domains for now, it's trivial to turn it back on if required.